cpp/include/cudf/detail/iterator.cuh

```diff
@@ -415,6 +621,164 @@ struct scalar_representation_pair_accessor : public scalar_value_accessor<Elemen
+ *
+ * @tparam Element The type of elements in the scalar
+ * @tparam mode The has_nulls mode of the iterator
+ * @tparam has_nulls If the scalar value will have a null at runtime
+ * @param scalar_value The scalar to iterate
+ * @return Iterator that returns scalar elements and validity of the
+ * element in a thrust::optional
+ */
+template <typename Element>
+auto inline make_optional_iterator(scalar const& scalar_value,
+                                   contains_nulls::DYNAMIC,
+                                   bool has_nulls)
+{
+  CUDF_EXPECTS(type_id_matches_device_storage_type<Element>(scalar_value.type().id()),
+               "the data type mismatch");
+  return thrust::make_transform_iterator(
+    thrust::make_constant_iterator<size_type>(0),
+    scalar_optional_accessor<Element, contains_nulls::DYNAMIC>{scalar_value, has_nulls});
+}
+
+/**
+ * @brief Constructs an optional iterator over a scalar's values and its validity.
+ *
+ * Dereferencing the returned iterator returns a `thrust::optional<Element>`.
+ *
+ * When the element of an iterator contextually converted to bool, the conversion returns true
+ * if the object contains a value and false if it does not contain a value.
+ *
+ * The iterator behavior is undefined if the scalar is destroyed before iterator dereferencing.
+ *
+ * make_optional_iterator ith mode `YES` means that the scalar supports nulls and
+ * potentially has null values, therefore the optional might not contain a value
+ * therefore the optional will always contain a value.
+ *
+ * Example:
+ *
+ * \code{.cpp}
+ * template<typename T, bool any_nulls>
+ * void some_function(cudf::column_view<T> const& col_view, scalar const& scalar_value){
+ *    auto d_col = cudf::column_device_view::create(col_view);
+ *    if constexpr(any_nulls) {
+ *      auto column_iterator = cudf::detail::make_optional_iterator<T>(d_col,
+ *                                                cudf::contains_nulls::YES{});
+ *      auto scalar_iterator = cudf::detail::make_optional_iterator<T>(scalar_value,
+ *                                                cudf::contains_nulls::YES{});
+ *      //use iterators
+ *    } else {
+ *      auto column_iterator = cudf::detail::make_optional_iterator<T>(d_col,
+ *                                                cudf::contains_nulls::NO{});
+ *      auto scalar_iterator = cudf::detail::make_optional_iterator<T>(scalar_value,
+ *                                                cudf::contains_nulls::NO{});
+ *      //use iterators
+ *    }
+ * }
+ * \endcode
+ *
+ * @throws cudf::logic_error if the scalar is not nullable, and `YES` mode used
+ * @throws cudf::logic_error if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar elements, and validity of the
+ * element in a thrust::optional
```

Comment:
```suggestion
 * @return Iterator that returns scalar elements and the validity of the
 * element in a thrust::optional
```

Reply:
Okay I corrected these, and removed documentation about a now removed template parameter.